### PR TITLE
[PLAY-1141] Prep icon kit for Power icons in React

### PR DESCRIPTION
**Pull request by markdoeswork**
**What does this PR do?** A clear and concise description with your runway ticket url.

Runway: https://nitro.powerhrg.com/runway/backlog_items/PLAY-1141

Nitro Alpha: https://pr37639.nitro-web.beta.gm.powerapp.cloud/user_requests/new?redirect_fallback=/sales/reps

The playbook icon library gives its consumers SVG files, I made the `icon` prop on the Icon kit `type: any` so it can accept svgs instead of just a string. Font Awesome applies the icon just based off the class name so this only needed to be a string before. 

The Rails version needed to have the assets in the asset pipeline because we wanted to call the with the same API `<%= pb_rails("icon" props: {icon: "roofing"})%>`. 

screenshot-nimbusweb me-2024 01 23-09_42_40

**How to test?** Steps to confirm the desired behavior:
1. Go to https://pr3058.playbook.beta.gm.powerapp.cloud/beta/icons
2. Witness the React playbook icon kit displaying playbook library's svgs.

1. Go to https://pr3058.playbook.beta.gm.powerapp.cloud/samples/icons
2. Witness the Rails playbook icon kit displaying playbook library's svgs. 

#### Checklist:
- [x] **LABELS** Add a label: `enhancement`, `bug`, `improvement`, `new kit`, `deprecated`, or `breaking`. See [Changelog & Labels](https://github.com/powerhome/playbook/wiki/Changelog-&-Labels) for details.
- [x] **DEPLOY** I have added the `milano` label to show I'm ready for a review.
- [😈] **TESTS** I have added test coverage to my code.

### CODE REVIEW

File: playbook/app/pb_kits/playbook/pb_icon/_icon.tsx [hunk 1]
```diff
@@ -27,7 +27,7 @@ type IconProps = {
   data?: {[key: string]: string},
   fixedWidth?: boolean,
   flip?: "horizontal" | "vertical" | "both" | "none",
-  icon: string,
+  icon: any,
```
Comment (markdoeswork) [inline playbook/app/pb_kits/playbook/pb_icon/_icon.tsx:30]: I couldn't get `icon: string | SVGElement`to work and I tried things like `Function` but typescript didn't like it

File: playbook/app/pb_kits/playbook/pb_icon/_icon.tsx [hunk 2]
```diff
@@ -123,6 +126,19 @@ const Icon = (props: IconProps) => {
           }
         </>
       )
+    else if (!iconString)
+      return (
+        <>
+          <span
+              {...dataProps}
+              {...htmlProps}
+              className="pbiconhere"
```
Comment (markdoeswork) [inline playbook/app/pb_kits/playbook/pb_icon/_icon.tsx:135]: This classname is kinda weak

File: playbook-website/app/javascript/components/Icons/index.tsx [hunk 3]
```diff
@@ -1,20 +1,19 @@
...
-import Calendar from "@powerhome/playbook-icons/icons/calendar.svg"
-import Filter from "@powerhome/playbook-icons/icons/filter.svg"
-import Edit from "@powerhome/playbook-icons/icons/edit.svg"
-import Trash from "@powerhome/playbook-icons/icons/trash.svg"
-import Check from "@powerhome/playbook-icons/icons/check.svg"
-import Plus from "@powerhome/playbook-icons/icons/plus.svg"
-import Search from "@powerhome/playbook-icons/icons/search.svg"
-
+import { Icon, Title, Flex, FlexItem, Card } from "playbook-ui"
+import Roofing from "!babel-loader!react-svg-loader!@powerhome/playbook-icons/icons/roofing.svg"
```
Comment (markdoeswork) [inline playbook-website/app/javascript/components/Icons/index.tsx:4]: I wanted to add this to the webpack config `playbook-website/config/webpack/environment.js` but i would need to specify the correct area to apply. And that was too hard

File: playbook/app/pb_kits/playbook/pb_icon/docs/_icon_custom.html.erb [hunk 4]
```diff
@@ -2,15 +2,9 @@
...
-  <p><%= pb_rails("icon", props: { rotation: 90, custom_icon: svg_url, size: "2x" } ) %></p>
-  <p><%= pb_rails("icon", props: { spin: true, custom_icon: svg_url, size: "3x" } ) %></p>
-  <p><%= pb_rails("icon", props: { size: "5x", custom_icon: svg_url } ) %></p>
-  <p><%= pb_rails("icon", props: { flip: "horizontal", size: "5x", custom_icon: svg_url } ) %></p>
-
-  <%= pb_rails("body", props: {
-    text: "Custom icons are compatible with other icon props (size, rotation,
-    spin, flip, etc). Their SVG fill colors will be inherited from
-    parent element's css color properties."
-  } ) %>
```
Comment (jasoncypret) [inline playbook/app/pb_kits/playbook/pb_icon/docs/_icon_custom.html.erb:15]: Why is this deleted? Are we not retaining these features?

File: playbook/app/pb_kits/playbook/pb_icon/docs/_icon_custom.md [hunk 5]
```diff
@@ -1,19 +1,14 @@
...
-When using custom icons it is important to introduce a "clean" SVG. In order to ensure these custom icons perform as intended within your kit(s), ensure these things have been modified from the original SVG markup:
-
-Attributes must be React compatible e.g. <code>xmlns:xlink</code> should be <code>xmlnsXlink</code> and so on. <strong>There should be no hyphenated attributes and no semi-colons!.</strong>
-
-Fill colors with regards to <code>g</code> or <code>path</code> nodes, e.g. <code>fill="black"</code>, should be replaced with   <code>currentColor</code> ala <code>fill="currentColor"</code>. Your mileage may vary depending on the complexity of your SVG.
-
-Pay attention to your custom icon's dimensions and `viewBox` attribute. It is best to use a `viewBox="0 0 512 512"` starting point __when designing instead of trying to retrofit the viewbox afterwards__!
-
-You must source *your own SVG into component/view* you are working on. This can easily be done in programmatic and maintainable ways.
-
```
Comment (jasoncypret) [inline playbook/app/pb_kits/playbook/pb_icon/docs/_icon_custom.md:12]: Isn't some of this still needed? 

`fill="black"`, should be replaced with `currentColor` ala `fill="currentColor"`

I believe we still need something like that to make colors respond to text size.

**Review by jasoncypret — CHANGES_REQUESTED** [review]
The icon is not responding to changes in font size for color.

File: playbook/app/pb_kits/playbook/pb_icon/docs/_icon_custom.md (hunk 5 above)
Comment (markdoeswork) [reply playbook/app/pb_kits/playbook/pb_icon/docs/_icon_custom.md:12]: this has been returned

File: playbook/app/pb_kits/playbook/pb_icon/docs/_icon_custom.html.erb (hunk 4 above)
Comment (markdoeswork) [reply playbook/app/pb_kits/playbook/pb_icon/docs/_icon_custom.html.erb:15]: this was also brought back

File: playbook/app/pb_kits/playbook/pb_icon/_icon.scss [hunk 6]
```diff
@@ -1,9 +1,51 @@
...
+      animation-timing-function: steps(8);
+    }
+    &.spin {
+      animation-name: pb_icon_spin;
+      animation-delay: 0s;
+      animation-direction: normal;
+      animation-duration: 2s;
+      animation-iteration-count: infinite;
+      animation-timing-function: linear;
+    }
```
Comment (jasoncypret) [inline playbook/app/pb_kits/playbook/pb_icon/_icon.scss:48]: I'd like to discuss these. We shouldn't have to add anything at the current stage to get the icons to work just like font awesome. Our custom icons have been obeying the added FA classes for years now.

File: playbook/app/pb_kits/playbook/pb_icon/_icon.tsx [hunk 7]
```diff
@@ -40,63 +42,119 @@ type IconProps = {
...
+  fa: {
+    90: 'fa-rotate-90',
+    180: 'fa-rotate-180',
+    270: 'fa-rotate-270'
+  },
+  svg: {
+    90: 'rotate_90',
+    180: 'rotate_180',
+    270: 'rotate_270'
+  }
```
Comment (jasoncypret) [inline playbook/app/pb_kits/playbook/pb_icon/_icon.tsx:98]: Why are we adding so many things that are already working in font awesome. The idea is to USE the FA stuff as is and extend our new icons to render as if they are coming from an FA library (Minus the FA Dash Initial Class)

File: playbook/app/pb_kits/playbook/pb_icon/icon.rb [hunk 8]
```diff
@@ -79,19 +79,42 @@ def custom_icon_classname
...
+          "2x" => 2,
+          "3x" => 3,
+          "4x" => 4,
+          "5x" => 5,
+          "6x" => 6,
+          "7x" => 7,
+          "8x" => 8,
+          "9x" => 9,
+          "10x" => 10 }
+      end
```
Comment (jasoncypret) [inline playbook/app/pb_kits/playbook/pb_icon/icon.rb:116]: I think this should work how it does now. Just allowing the classes and let our fa.css size this correctly.

File: playbook-website/app/views/samples/icons.html.erb [hunk 9]
```diff
@@ -0,0 +1,172 @@
...
+  .blue_color {
+    color: blue
+  }
+  .red_color {
+    color: red
+  }
+  .green_color {
+    color: green
+  }
+</style>
```
Comment (jasoncypret) [inline playbook-website/app/views/samples/icons.html.erb:172]:
```suggestion
```

File: playbook/app/pb_kits/playbook/pb_icon/_icon.scss [hunk 10]
```diff
@@ -1,9 +1,48 @@
...
```
[the remark this answers is not in the document]
Comment (markdoeswork) [reply playbook/app/pb_kits/playbook/pb_icon/_icon.scss]: This has been all removed, using the FA styles works

File: playbook-website/app/views/samples/icons.html.erb [hunk 11]
```diff
@@ -0,0 +1,172 @@
...
+<%= pb_rails("flex", props: { padding_left: "md"} ) do %>
+  <%= pb_rails("flex/flex_item") do %>
+    <%= pb_rails("title", props: {size: 3, text: "Power Icons", padding_bottom: "sm"})%>
+  <% end %>
+<% end %>
+
+<%= pb_rails("flex", props: { padding_left: "md" }) do %>
+  <%= pb_rails("card", props: { padding: "none", header: true, margin_bottom: "sm"}) do %>
+    <%= pb_rails("card/card_body", props: { padding: "md",  }) do %>
+      <%= pb_rails("icon", props: { icon: image_pack_url("roofing.svg") }) %>
```
[the remark this answers is not in the document]
Comment (markdoeswork) [reply playbook-website/app/views/samples/icons.html.erb]: I've put the svgs in the asset pipeline so that we can achieve the suggested prop style
 `<%= pb_rails("icon", props: { icon: "roofing" }) %>` works now

File: playbook/app/pb_kits/playbook/pb_icon/_icon.scss (hunk 6 above)
Comment (markdoeswork) [reply playbook/app/pb_kits/playbook/pb_icon/_icon.scss]: This has been all removed, using the FA styles works

File: playbook/app/pb_kits/playbook/pb_icon/icon.rb (hunk 8 above)
Comment (markdoeswork) [reply playbook/app/pb_kits/playbook/pb_icon/icon.rb]: This is removed, we just need to set the `width="auto"` and `height="auto"` on the svg to get the FA styles to apply size correctly

File: playbook-website/app/views/samples/icons.html.erb (hunk 9 above)
Comment (markdoeswork) [reply playbook-website/app/views/samples/icons.html.erb]: This has been all removed, wrapping the icon kit in body kit can change their color

File: playbook/app/pb_kits/playbook/pb_icon/_icon.tsx (hunk 7 above)
Comment (markdoeswork) [reply playbook/app/pb_kits/playbook/pb_icon/_icon.tsx]: This has been all removed, using the FA styles works

File: playbook/app/pb_kits/playbook/pb_icon/_icon.tsx [hunk 12]
```diff
@@ -40,63 +42,119 @@ type IconProps = {
   spin?: boolean,
 } & GlobalProps
 
+const iconSizeMap = {
+  base: 16,
+  xs: .75,
+  sm: .875,
+  lg: 1.25
+}
```
[the remark this answers is not in the document]
Comment (markdoeswork) [reply playbook/app/pb_kits/playbook/pb_icon/_icon.tsx]: This has been removed, FA styles handle this

File: playbook-website/app/javascript/components/Website/src/pages/IconList/index.tsx [hunk 13]
```diff
@@ -1,27 +1,32 @@
...
+import ChevronDown from "!file-loader!@powerhome/playbook-icons/icons/chevron-down.svg"
+import Times from "!file-loader!@powerhome/playbook-icons/icons/times.svg"
+import Bars from "!file-loader!@powerhome/playbook-icons/icons/bars.svg"
+import Calendar from "!file-loader!@powerhome/playbook-icons/icons/calendar.svg"
+import Filter from "!file-loader!@powerhome/playbook-icons/icons/filter.svg"
+import Edit from "!file-loader!@powerhome/playbook-icons/icons/edit.svg"
+import Trash from "!file-loader!@powerhome/playbook-icons/icons/trash.svg"
+import Check from "!file-loader!@powerhome/playbook-icons/icons/check.svg"
+import Plus from "!file-loader!@powerhome/playbook-icons/icons/plus.svg"
+import Search from "!file-loader!@powerhome/playbook-icons/icons/search.svg"
```
[the remark this answers is not in the document]
Comment (markdoeswork) [reply playbook-website/app/javascript/components/Website/src/pages/IconList/index.tsx]: I thought we weren't going to use the `playbook-icons-react`, it works it just need to have `@powerhome` prefix `import { Roofing, Powergon, Nitro, ChevronDown, Times, Bars, Calendar, Filter, Edit, Trash, Check, Plus, Search} from '@powerhome/playbook-icons-react'`

File: playbook/app/pb_kits/playbook/pb_icon/docs/_icon_custom.md [hunk 14]
```diff
@@ -4,16 +4,21 @@ When using custom icons it is important to introduce a "clean" SVG. In order to
 
 Attributes must be React compatible e.g. <code>xmlns:xlink</code> should be <code>xmlnsXlink</code> and so on. <strong>There should be no hyphenated attributes and no semi-colons!.</strong>
 
-Fill colors with regards to <code>g</code> or <code>path</code> nodes, e.g. <code>fill="black"</code>, should be replaced with   <code>currentColor</code> ala <code>fill="currentColor"</code>. Your mileage may vary depending on the complexity of your SVG.
+Fill colors with regards to <code>g</code> or <code>path</code> nodes, e.g. <code>fill="black"</code>, should be replaced with <code>currentColor</code> ala <code>fill="currentColor"</code>. Your mileage may vary depending on the complexity of your SVG.
```
Comment (jasoncypret) [inline playbook/app/pb_kits/playbook/pb_icon/docs/_icon_custom.md:7]: It would be awesome to build a page or a script where a user could fix any problems with their SVG's

File: playbook/app/pb_kits/playbook/pb_icon/docs/_icon_custom.md [hunk 15]
```diff
@@ -4,16 +4,21 @@ When using custom icons it is important to introduce a "clean" SVG. In order to
...
 
 ### React
 
-So long as you have a valid React `<SVG>` node, you can send it as the `customIcon` prop and the kit will take care of the rest.
+- Providing a valid React `<SVG>` element to the `icon` prop results in an `<SVG>` node within the working view.
+- Sending the absolute path to your SVG (e.g. `/my/path/to/icon.svg`) results in an `img` node with the `src` attribute set to the provided path:
+
+  ```html
+  <img src="host.com/my/path/to/icon.svg" />
+  ```
```
[the remark this answers is not in the document]
Comment (markdoeswork) [reply playbook/app/pb_kits/playbook/pb_icon/docs/_icon_custom.md:20]: Stephen just put this in here to prove that it would work ima remove it

File: playbook/app/pb_kits/playbook/pb_icon/_icon.tsx [hunk 16]
```diff
@@ -136,7 +136,19 @@ const Icon = (props: IconProps) => {
...
+    else if (iconURL)
+      return (
+        <>
+          <span
+              {...dataProps}
+              {...htmlProps}
+              className={classesEmoji}
+              id={id}
+          >
+            <img src={iconURL} />
```
[the remark this answers is not in the document]
Comment (markdoeswork) [reply playbook/app/pb_kits/playbook/pb_icon/_icon.tsx:148]: this has been removed

File: playbook-website/app/javascript/components/Website/src/pages/IconList/index.tsx [hunk 17]
```diff
@@ -35,32 +34,135 @@ export default function IconList() {
...
+              cursor='pointer'
+            >
+              <Icon icon={<ChevronDown />} />
+            </Card>
+            <Card
+              marginRight='sm'
+              hover={{ shadow: "deeper" }}
+              cursor='pointer'
+            >
+              <Icon icon={<Times />} />
```
[the remark this answers is not in the document]
Comment (markdoeswork) [reply playbook-website/app/javascript/components/Website/src/pages/IconList/index.tsx:87]: There needs to be a place where `playbook-ui` and `playbook-icons` shake hands. Doing it every single time `<Icon/>` is called in Nitro is for sure a bad idea. A big issues is the "don't do `playbook-icons` in Playbook" constraint

[the remark this answers is not in the document]
Comment (thestephenmarshall) [reply playbook-website/app/javascript/components/Website/src/pages/IconList/index.tsx:87]: So, Nitro can provide an optional configuration object. Consider this non-working pseudo-code:

```tsx
// this lives globally in Nitro
// determined by Nitro's build system
type PowerIcons = {[key: string]: ReactSVGElement}
const POWER_ICONS: PowerIcons = window.POWER_ICONS
const powerIcon = typeof(window.POWER_ICONS) == "object" ? window.POWER_ICONS[icon] : undefined
```

This bit of code just provides a window level configuration variable which Playbook can optionally read. IF it is present and IF we have say `times` as a key in the object THEN the kit can simply use the SVG. Best part is Playbook does not need to require our `power-icons`, but can still use them.

[the remark this answers is not in the document]
Comment (thestephenmarshall) [reply playbook-website/app/javascript/components/Website/src/pages/IconList/index.tsx:87]: Please review: https://github.com/powerhome/playbook/pull/3058/commits/6734fdbaf30e601c46a611a37868ded9da172adb

[the remark this answers is not in the document]
Comment (thestephenmarshall) [reply playbook-website/app/javascript/components/Website/src/pages/IconList/index.tsx:87]: Here is a playground that proves this works which is using this alpha: https://pr788.playgrounds.beta.gm.powerapp.cloud/

PR: https://github.com/powerhome/playgrounds/pull/788

File: Dockerfile [hunk 18]
```diff
@@ -80,3 +80,6 @@ COPY --link --chown=9999:9999 playbook /home/app/src/playbook
 COPY --link --chown=9999:9999 playbook-website /home/app/src/playbook-website
 COPY --link --from=release /home/app/src/playbook/dist /home/app/src/playbook/dist
 COPY --link --from=release /home/app/src/playbook-website/public /home/app/src/playbook-website/public
+COPY --link --from=release /home/app/src/node_modules/@powerhome/playbook-icons/icons /home/app/src/temp-icons
```
Comment (markdoeswork) [inline Dockerfile:83]: Force review env and production to insert icons in asset pipeline

File: playbook-website/app/javascript/components/Website/src/pages/IconList/index.tsx (hunk 17 above)
[the remark this answers is not in the document]
Comment (markdoeswork) [reply playbook-website/app/javascript/components/Website/src/pages/IconList/index.tsx]: I import the icons onto the example page like this 

```
import { Roofing, Powergon, Nitro, ChevronDown, Times, Bars, Calendar, Filter, Edit, Trash, Check, Plus, Search} from '@powerhome/playbook-icons-react'

const pbIcons = {
  roofing: Roofing,
  nitro: Nitro,
  powergon: Powergon,
  chevrondown: ChevronDown,
  times: Times,
  bars: Bars,
  calendar: Calendar,
  filter: Filter,
  edit: Edit,
  trash: Trash,
  check: Check,
  plus: Plus,
  search: Search
}

window.PB_ICONS = pbIcons
```

And the react icon kit listens for `PB_ICONS` and loads them if they are there

**Review by thestephenmarshall — APPROVED** [review]

**Review by jasoncypret — APPROVED** [review]

**Review by jasperfurniss — APPROVED** [review]
